### PR TITLE
At - Index typeclass

### PR DESCRIPTION
**Pull request by julien-truffaut**
- I used the same encoding as for `FieldX` and `Cons`.
- Introduced inheritance between `At` and `Index`
- I wonder if we should have an `AtK` version for higher kinded `S`, e.g. `Atk[List, Int]` instead of `At[List[A], Int, A]`

### CODE REVIEW

File: kernel/shared/src/main/scala/monocle/Optional.scala [hunk 1]
```diff
@@ -15,6 +15,7 @@ trait Optional[A, B] { self =>
   }
 
   def composeLens[C](other: Lens[B, C]): Optional[A, C] = compose(other)
+  def composePrism[C](other: Prism[B, C]): Optional[A, C] = compose(other)
```
Comment (cquiroz) [inline kernel/shared/src/main/scala/monocle/Optional.scala]: Are you planning to keep the symbolic operators?

File: kernel/shared/src/main/scala/monocle/function/Index.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
+package monocle.function
+
+import monocle.Optional
+
+import scala.util.Try
+
+trait Index[S] {
```
Comment (cquiroz) [inline kernel/shared/src/main/scala/monocle/function/Index.scala]: So it is a value of type `A` indxed by  a type `I`? I guess that would be applicable to list like and map like instances 👍

File: kernel/shared/src/main/scala/monocle/Optional.scala (hunk 1 above)
Comment (julien-truffaut) [reply kernel/shared/src/main/scala/monocle/Optional.scala]: I don't know. Maybe not in kernel but we can always have them defined in an extra module. What do you think? In scala 3, we should only have one compose function.

File: kernel/shared/src/main/scala/monocle/function/Index.scala (hunk 2 above)
Comment (julien-truffaut) [reply kernel/shared/src/main/scala/monocle/function/Index.scala]: indeed, it is even applicable to a String

**Pull request merged**